### DISCUSSION TITLE
TypeError: issubclass() arg 1 must be a class with BaseModel

### DISCUSSION THREAD

**Discussion by farnsworth** [pydantic/pydantic#5970] [Question] [answered]
I would like to check if the signature of a function has arguments inherited from a pydantic BaseModel. To do that I use `issubclass(typ, BaseModel)` where `typ` is the annotation of the function argument, but for instance:
```
from pydantic import BaseModel
issubclass(list[BaseModel], BaseModel)
```
gives:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.10/abc.py", line 123, in __subclasscheck__
    return _abc_subclasscheck(cls, subclass)
TypeError: issubclass() arg 1 must be a class
```
but actually `isinstance(list[BaseModel], type)` is `True` and I'm not able to understand the error.

**Comment by dmontagu** [comment, marked as answer]
This is caused by the fact that `BaseModel` is an `ABC` (specifically, its metaclass inherits from `abc.ABCMeta`).

As a result, BaseModel has special `issubclass` logic, and when you try to do `issubclass(<something>, BaseModel)`, it seems somewhere it's passing `<something>` as arg 1 of `issubclass`. Considering you don't get this error for non-`ABC` types, might be considered a bug in `abc` but either way it's not surprising because even though `isinstance(list[BaseModel], type)` may return `True`, `list[BaseModel]` (or `list[<anything>]` for that matter) isn't _really_ a type, it's an instance of `types.GenericAlias`:
```python
from pydantic import BaseModel

print(type(list[BaseModel]))
#> types.GenericAlias
```
In most scenarios it behaves enough like a type that this doesn't cause issues, but apparently it doesn't work with ABC-subclass-checks.

I suspect there is a better way to structure your code to avoid this particular check, but you might also ask on CPython or similar whether the following code (which doesn't involve pydantic at all) should be considered a bug:
```python
from abc import ABC

class A(ABC):
    pass

issubclass(list[A], A)
#> TypeError: issubclass() arg 1 must be a class
```
